### PR TITLE
Redesign testing fixtures

### LINKED ISSUE #250: Serial integration tests end with segfault (though successful)
There is sth wrong with the clean-up of the integration tests. 
When running only the serial integration tests:
```
./tools/compileAndTest.py -t --run_test="PreciceTests/Serial/*"
```
I get a:
```
[1]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
```
but only for two out of the four ranks. 

When running all integration tests, this disappears:
```
./tools/compileAndTest.py -t --run_test="PreciceTests/*/*"
```

### LINKED ISSUE #399: Write unit tests for `com/Communication.cpp`
Code coverage tells us that the advanced methods, such as broadcast etc. are not tested at all. Those basic implementations are used in the `SocketsCommunication`.

### LINKED ISSUE #597: Refactor initialization and finalization of tests
We currently start all our tests from a single entry point in `src/testing/main.cpp` with some things initialized and finalized there:

https://github.com/precice/precice/blob/a5bc1e7652ed957d817c9fcb46848fd760a474fc/src/testing/main.cpp#L98-L131

These initialization and finalization efforts however make not sense for every test, e.g. in an integration test, the `EventRegistry` should be initialized with the `GlobalCommunicator` of the participant (so through the preCICE API), not with the `GlobalCommunicator` of the whole tests. 

For other tests, certain initialization is also missing: e.g. for integration tests in `PreciceTests/Serial/*`, we don't restrict the MPI communicator as we do for integration tests in `PreciceTests/Parallel/*`.

https://github.com/precice/precice/blob/a5bc1e7652ed957d817c9fcb46848fd760a474fc/src/testing/Fixtures.hpp#L81-L99

We currently work around this issues with several (historically grown) hacks, explicit setup of environments directly in the tests and prohibiting certain features for certain tests (e.g. PETSc based RBF in serial integration tests). Furthermore, our tests are sensitive to their order, which makes them very hard (thus time consuming) to maintain. 

Let's do a clean separation of different testcases and their initialization / finalization and a clean definition which things we initialize / finalize and which things we treat with `precice::testMode`. Either through different entry points or through fixtures alone. We need to distinguish the following cases:
* **(US)** serial unit tests, example: `MappingTests/NearestNeighborMapping`
* **(UP)** parallel unit tests on a single participant, example: `MappingTests/PetRadialBasisFunctionMapping/Parallel`
* **(UMS)** unit tests with two or more serial participants, example: `CplSchemeTests/SerialImplicitCouplingSchemeTests/*`
* **(UMP)** unit tests with multiple participants and at least one parallel participant, example: `PartitionTests/*`  
* **(IS)** integration tests with one, two or more serial participants, example: `PreciceTests/Serial/*`
* **(IP)** integration tests with multiple participants and at least one parallel participant, example: `PreciceTests/Parallel/*`

### PR DESCRIPTION
**Pull request by fsimonis**
This PR is the complete redesign of the testing fixtures to cover all the cases described in #597.
The primary goal is to generate clean and consistent environments for each test. There should be close to no inter-test dependencies.

## Goals
The secondary goals of this PR are:
1. to simplify the notation of tests. Especially the notation of integration tests were highly obscure due to the required setup.
2. to reduce the setup and error checks to be done by the developer.

## Changes
To achieve said goals, I had to:
* Introduce a custom notation for tests that cover integration tests as well as all unit test cases described in #597.
* Redesign the way preCICE handles MPI communicators. This lead to a complete overhaul of `utils/Parallel.h`. It now uses shared linked `CommStates` that free communicators when required.
* Redesign of `MPIDirectCommunciation`, which is now only usable for participants inside a shared communicator. This required a slight modification of `com::Communication`. I also added tests for all communication routines and cases.
* Add a TestContext helper type which exposes all required information to a running test (Name, rank, size, comm) and allows to connect participants.
* Remove the majority of test fixtures and decorators as they are now super-seeded by the TestContext.
* Port and fix all tests.

Additional decisions reflected in this PR:
* Some tests one setup with different configurations. I broke the tests down into tests running one configuration each. This simplifies debugging and improves error messages.
* Tests now require to be run on 4 MPI Ranks and will be disabled when configured without `PRECICE_MPICommunication`. Running tests on less than 4 ranks covers a miniscule part of the library. The tests will still compile without MPI, thus testing compile-time information.
* The solverdummies (c, cpp, fortran) are now build against precice as additional tests. Further test run coupling cases with the solverdummies. This ensures that we don't break the solverdummies and provides some level of testing for preCICE build without MPI.
* Ranks in tests now synchronize at the beginning of the setup, at the end of the setup (individual participants) and at the end of the test. This slows down tests, but improves the resulting logs.

The notation for tests looks as follows:

```cpp
// Integration test
BOOST_AUTO_TEST_CASE(LocalRBFPartitioning)
{
  PRECICE_TEST("SolverOne"_on(3_ranks), "SolverTwo"_on(1_rank));
  // ...
  SolverInterface interface(context.name, configFilename, context.rank, context.size);
  // ...
  if (context.name == "SolverOne") { //...
    if (context.isMaster()) {
      //...
    } else if (context.isRank(1)) {
      //...
    } else {
      //...
    }
  }
}

// Unit tests with an unnamed participant on 4 ranks with MS setup and initialized PETSc (which implies Events),
BOOST_AUTO_TEST_CASE(DistributedConsistent2DV1)
{
  PRECICE_TEST(""_on(4_ranks).setupMasterSlaves(), Require::PETSc);
}
```

## Test Suite
The full test suite looks as follows:
```
Test project /tmp/precice
      Start  1: precice.acceleration
 1/29 Test  #1: precice.acceleration ......................   Passed    0.14 sec
      Start  2: precice.action
 2/29 Test  #2: precice.action ............................   Passed    0.64 sec
      Start  3: precice.com
 3/29 Test  #3: precice.com ...............................   Passed    0.83 sec
      Start  4: precice.com.mpiports
 4/29 Test  #4: precice.com.mpiports ......................   Passed    0.12 sec
      Start  5: precice.cplscheme
 5/29 Test  #5: precice.cplscheme .........................   Passed    4.37 sec
      Start  6: precice.io
 6/29 Test  #6: precice.io ................................   Passed    0.10 sec
      Start  7: precice.m2n
 7/29 Test  #7: precice.m2n ...............................   Passed    0.19 sec
      Start  8: precice.m2n.mpiports
 8/29 Test  #8: precice.m2n.mpiports ......................   Passed    0.09 sec
      Start  9: precice.mapping
 9/29 Test  #9: precice.mapping ...........................   Passed    0.22 sec
      Start 10: precice.mapping.petrbf
10/29 Test #10: precice.mapping.petrbf ....................   Passed    0.94 sec
      Start 11: precice.math
11/29 Test #11: precice.math ..............................   Passed    0.07 sec
      Start 12: precice.mesh
12/29 Test #12: precice.mesh ..............................   Passed    0.08 sec
      Start 13: precice.partition
13/29 Test #13: precice.partition .........................   Passed    0.55 sec
      Start 14: precice.interface
14/29 Test #14: precice.interface .........................   Passed    0.05 sec
      Start 15: precice.serial
15/29 Test #15: precice.serial ............................   Passed    6.68 sec
      Start 16: precice.parallel
16/29 Test #16: precice.parallel ..........................***Failed    1.40 sec
      Start 17: precice.query
17/29 Test #17: precice.query .............................   Passed    0.10 sec
      Start 18: precice.testing
18/29 Test #18: precice.testing ...........................   Passed    0.07 sec
      Start 19: precice.utils
19/29 Test #19: precice.utils .............................   Passed    0.06 sec
      Start 20: precice.xml
20/29 Test #20: precice.xml ...............................   Passed    0.29 sec
21/29 Test #21: precice.solverdummy.build.cpp .............   Passed    0.47 sec
      Start 22: precice.solverdummy.build.c
22/29 Test #22: precice.solverdummy.build.c ...............   Passed    0.22 sec
      Start 23: precice.solverdummy.build.fortran
23/29 Test #23: precice.solverdummy.build.fortran .........   Passed    0.25 sec
      Start 24: precice.solverdummy.run.cpp-cpp
24/29 Test #24: precice.solverdummy.run.cpp-cpp ...........   Passed    0.34 sec
      Start 25: precice.solverdummy.run.c-c
25/29 Test #25: precice.solverdummy.run.c-c ...............   Passed    0.39 sec
      Start 26: precice.solverdummy.run.fortran-fortran
26/29 Test #26: precice.solverdummy.run.fortran-fortran ...   Passed    0.39 sec
      Start 27: precice.solverdummy.run.cpp-c
27/29 Test #27: precice.solverdummy.run.cpp-c .............   Passed    0.40 sec
      Start 28: precice.solverdummy.run.cpp-fortran
28/29 Test #28: precice.solverdummy.run.cpp-fortran .......   Passed    0.38 sec
      Start 29: precice.solverdummy.run.c-fortran
29/29 Test #29: precice.solverdummy.run.c-fortran .........   Passed    0.37 sec

97% tests passed, 1 tests failed out of 29

Label Time Summary:
Solverdummy    =   3.22 sec*proc (9 tests)
canfail        =   0.21 sec*proc (2 tests)
mpiports       =   0.21 sec*proc (2 tests)
petsc          =   0.94 sec*proc (1 test)

Total Test time (real) =  20.33 sec

The following tests FAILED:
	 16 - precice.parallel (Failed)
Errors while running CTest
```
_Note: @uekerman is redesigning the failing QN tests._

Closes #597 
Closes #250 
Closes #399

### CODE REVIEW

File: src/testing/TestContext.hpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+  /// the name of the pariticipant
+  std::string name;
+
+  /// the amount of ranks this participant runs on
+  int         size   = 1;
+
+  /// wheather to initialize a master-slave communication for this participant
+  bool        initMS = false;
+
+  /// Constructs a serial pariticipant with a given name
```
Comment (uekerman) [inline src/testing/TestContext.hpp]:
```suggestion
  /// Constructs a serial participant with a given name
```

**Comment by MakisH** [comment]
Test results on Ubuntu 20.04 (daily build 30/3/2020), 3a7210bc7f97859f71bf5ac2a62a4a9e4edf6ff7:

```
Test project /home/tester/precice/build
      Start  1: precice.acceleration
 1/29 Test  #1: precice.acceleration ......................   Passed    1.11 sec
      Start  2: precice.action
 2/29 Test  #2: precice.action ............................   Passed    1.11 sec
      Start  3: precice.com
 3/29 Test  #3: precice.com ...............................   Passed    1.59 sec
      Start  4: precice.com.mpiports
 4/29 Test  #4: precice.com.mpiports ......................***Timeout  20.11 sec
      Start  5: precice.cplscheme
 5/29 Test  #5: precice.cplscheme .........................   Passed    3.01 sec
      Start  6: precice.io
 6/29 Test  #6: precice.io ................................   Passed    0.80 sec
      Start  7: precice.m2n
 7/29 Test  #7: precice.m2n ...............................   Passed    1.52 sec
      Start  8: precice.m2n.mpiports
 8/29 Test  #8: precice.m2n.mpiports ......................   Passed    1.38 sec
      Start  9: precice.mapping
 9/29 Test  #9: precice.mapping ...........................   Passed    1.03 sec
      Start 10: precice.mapping.petrbf
10/29 Test #10: precice.mapping.petrbf ....................   Passed    4.37 sec
      Start 11: precice.math
11/29 Test #11: precice.math ..............................   Passed    0.60 sec
      Start 12: precice.mesh
12/29 Test #12: precice.mesh ..............................   Passed    0.79 sec
      Start 13: precice.partition
13/29 Test #13: precice.partition .........................   Passed    2.52 sec
      Start 14: precice.interface
14/29 Test #14: precice.interface .........................   Passed    0.54 sec
      Start 15: precice.serial
15/29 Test #15: precice.serial ............................   Passed    5.97 sec
      Start 16: precice.parallel
16/29 Test #16: precice.parallel ..........................   Passed    6.32 sec
      Start 17: precice.query
17/29 Test #17: precice.query .............................   Passed    0.65 sec
      Start 18: precice.testing
18/29 Test #18: precice.testing ...........................   Passed    1.03 sec
      Start 19: precice.utils
19/29 Test #19: precice.utils .............................   Passed    1.18 sec
      Start 20: precice.xml
20/29 Test #20: precice.xml ...............................   Passed    1.25 sec
      Start 21: precice.solverdummy.build.cpp
21/29 Test #21: precice.solverdummy.build.cpp .............   Passed    1.54 sec
      Start 22: precice.solverdummy.build.c
22/29 Test #22: precice.solverdummy.build.c ...............   Passed    1.04 sec
      Start 23: precice.solverdummy.build.fortran
23/29 Test #23: precice.solverdummy.build.fortran .........   Passed    1.47 sec
      Start 24: precice.solverdummy.run.cpp-cpp
24/29 Test #24: precice.solverdummy.run.cpp-cpp ...........***Failed    0.73 sec
      Start 25: precice.solverdummy.run.c-c
25/29 Test #25: precice.solverdummy.run.c-c ...............   Passed    0.63 sec
      Start 26: precice.solverdummy.run.fortran-fortran
26/29 Test #26: precice.solverdummy.run.fortran-fortran ...   Passed    0.70 sec
      Start 27: precice.solverdummy.run.cpp-c
27/29 Test #27: precice.solverdummy.run.cpp-c .............   Passed    0.75 sec
      Start 28: precice.solverdummy.run.cpp-fortran
28/29 Test #28: precice.solverdummy.run.cpp-fortran .......   Passed    0.91 sec
      Start 29: precice.solverdummy.run.c-fortran
29/29 Test #29: precice.solverdummy.run.c-fortran .........   Passed    0.82 sec

93% tests passed, 2 tests failed out of 29

Label Time Summary:
Solverdummy    =   8.59 sec*proc (9 tests)
canfail        =  21.49 sec*proc (2 tests)
mpiports       =  21.49 sec*proc (2 tests)
petsc          =   4.37 sec*proc (1 test)

Total Test time (real) =  65.49 sec

The following tests FAILED:
	  4 - precice.com.mpiports (Timeout)
	 24 - precice.solverdummy.run.cpp-cpp (Failed)
Errors while running CTest
make: *** [Makefile:130: test] Error 8
```

Interestingly, running the tests a second time I get a success for `precice.solverdummy.run.cpp-cpp`, but this is not robust. Sometimes it fails, sometimes it succeeds. I ran the tests 6 times, I got a failure 2 times.

File: src/partition/ProvidedPartition.cpp [hunk 2]
```diff
@@ -137,14 +137,18 @@ void ProvidedPartition::prepare()
...
       if (utils::MasterSlave::isMaster()) {
         PRECICE_DEBUG("Fill vertex distribution");
+        auto & localIds = _mesh->getVertexDistribution()[0];
         for (int i = 0; i < _mesh->getVertexOffsets()[0]; i++) {
-          _mesh->getVertexDistribution()[0].push_back(i);
+          localIds.push_back(i);
         }
         for (int rankSlave = 1; rankSlave < utils::MasterSlave::getSize(); rankSlave++) {
+          // This always creates an entry for each slave
+          auto & remoteIds = _mesh->getVertexDistribution()[rankSlave];
```
Comment (uekerman) [inline src/partition/ProvidedPartition.cpp]:
```suggestion
          auto & slaveIds = _mesh->getVertexDistribution()[rankSlave];
```
With "remote", we typically refer to the other participant.

File: src/partition/ProvidedPartition.cpp [hunk 3]
```diff
@@ -137,14 +137,18 @@ void ProvidedPartition::prepare()
...
           for (int i = _mesh->getVertexOffsets()[rankSlave - 1]; i < _mesh->getVertexOffsets()[rankSlave]; i++) {
-            _mesh->getVertexDistribution()[rankSlave].push_back(i);
+            remoteIds.push_back(i);
```
Comment (uekerman) [inline src/partition/ProvidedPartition.cpp]:
```suggestion
            slaveIds.push_back(i);
```

File: src/testing/Fixtures.hpp [hunk 4]
```diff
@@ -12,94 +12,5 @@
 namespace precice {
 namespace testing {
```
Comment (uekerman) [inline src/testing/Fixtures.hpp]: File can be deleted.

File: src/testing/TestContext.hpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  /** Creates a M2N and establishes a master-master connection between participants
+   * @param[in] acceptor the accepting Pariticipant
+   * @param[in] requestor the requesting Pariticipant
+   * @param[in] options a set of options concerning the created connection
+   * 
+   * @note This function throws if the acceptor or requestor are unknown!
+   *
+   * @see ConnectionOptions
+   */
+  m2n::PtrM2N connect(const std::string &acceptor, const std::string &requestor, const ConnectionOptions &options = ConnectionOptions{}) const;
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/testing/TestContext.hpp]: You are right. Being explicit here seems to be the better fit.

**Comment by fsimonis** [comment]
@MakisH Maybe stale connection files?
Can you run `find -type d -name "precice-run" -exec rm -fr {} \;` and try again?
From the logs:

```
--------------------------------------------------------------------------
The user has called an operation involving MPI_Connect and/or MPI_Accept
that spans multiple invocations of mpirun. This requires the support of
the ompi-server tool, which must be executing somewhere that can be
accessed by all participants.

Please ensure the tool is running, and provide each mpirun with the MCA
parameter "pmix_server_uri" pointing to it.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
Your application has invoked an MPI function that is not supported in
this environment.

  MPI function: MPI_Comm_connect
  Reason:       Underlying runtime environment does not support accept/connect functionality
--------------------------------------------------------------------------
```

The `spans multiple invocations of mpirun` is a bit fishy.

**Comment by MakisH** [comment]
I pulled the latest state and it seems to run smoothly (apart from the `precice.com.mpiports` that always fails). Here is the output:

[tests_2004_2.log](https://github.com/precice/precice/files/4416653/tests_2004_2.log)

**Comment by uekerman** [comment]
> Maybe stale connection files?

I don't think so. I had a similar problem and made sure that the connection files are always removed. Sounds a bit like we are missing a synchronization after each solverdummy test? Or does cmake even start two solverdummy tests simultaneously?

**Comment by fsimonis** [comment]
> Or does cmake even start two solverdummy tests simultaneously?

I set the the [`RUN_SERIAL`](https://cmake.org/cmake/help/v3.10/prop_test/RUN_SERIAL.html) property of every preCICE test. So this should never be the case.

Tests are run in sequence.

**Comment by fsimonis** [comment]
The solverdummy tests also sometimes fail on travis. This is weird, but definitely a good hint that we should investigate this problem.

**Comment by uekerman** [comment]
Running them "manually" I cannot reproduce the problem. So it is related to the interplay "ctest <-> solverdummies".

**Comment by fsimonis** [comment]
I managed to fix the solverdummy problem. Parallel execution is not possible as we try to use it in CMake. I had to add an execution wrapper.

**Comment by fsimonis** [comment]
This PR is now ready for final review and merge.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,14 @@ All notable changes to this project will be documented in this file. For future
 
 
 - Implemented a new action to sum up data values, `action:summation`.
+- Added many tests for the communication abstraction.
+- Refactor `com::Communication` handling of rank adjustments.
+- Changed test setup to a simpler and consistent version using the new `testing::TestContext`. Tests now require to run on 4 MPI ranks. They will still compile but not run when `MPICommunication=OFF`.
+- Added tests which build solverdummies and test which run them in various configurations.
+- Enable RBF-based tests in partiton unit-tests and serial integration tests
+- Split multi-setup intergration tests into multiple single-setup tests.
```
Comment (uekerman) [inline CHANGELOG.md:12]:
```suggestion
- Split multi-setup integration tests into multiple single-setup tests.
```

**Pull request merged**